### PR TITLE
Add a feature to toggle between panes in the core #10555

### PR DESCRIPTION
**Pull request by arthur801031**
- I'm not sure if I need to add a menu item for this feature inside "Navigate".
- I'm also not sure if I placed the code inside the correct files since I'm new to open source project contributions. 

Thank you!

### CODE REVIEW

**Comment by swmitra** [comment]
@arthur801031 Can you please add/update some tests in MainViewManager test file?

File: src/nls/root/strings.js [hunk 1]
```diff
@@ -418,6 +418,7 @@ define({
     "CMD_SHOW_IN_EXPLORER"                : "Show in Explorer",
     "CMD_SHOW_IN_FINDER"                  : "Show in Finder",
     "CMD_SHOW_IN_OS"                      : "Show in OS",
+    "CMD_SPLITVIEW_TOGGLE"                : "Toggle Panes",
```
Comment (swmitra) [inline src/nls/root/strings.js]: As this command is for shuffling focus across panes, can we name it accordingly?

File: src/view/Pane.js [hunk 2]
```diff
@@ -209,6 +210,19 @@ define(function (require, exports, module) {
     }
 
     /**
+     * Toggle between panes
+     */
+    function _handleSwitchPane() {
```
Comment (swmitra) [inline src/view/Pane.js]: This entire logic of shuffling focus across panes could be in MainViewManager. Pane should contain operations specific to creation, state mutation and handling other things inside a Pane. As this operation is a user requested focus preemption, it could be at a higher level above Panes.

File: src/nls/root/strings.js (hunk 1 above)
Comment (arthur801031) [inline src/nls/root/strings.js]: Can we write it like this?
`"CMD_SHUFFLE_FOCUS_ACROSS_PANES"           : "Shuffle Focus Across Panes",`

File: src/view/MainViewManager.js [hunk 3]
```diff
@@ -1658,8 +1658,8 @@ define(function (require, exports, module) {
 
         return result;
     }
-
-
+
```
Comment (swmitra) [inline src/view/MainViewManager.js]: Please revert this change as it's only unintentional whitespace modification.

File: src/view/Pane.js (hunk 2 above)
Comment (arthur801031) [inline src/view/Pane.js]: I will try moving this function to MainViewManager. Thank you!

**Comment by arthur801031** [comment]
Hi @swmitra @petetnt , I'm having trouble using the Jasmine Spec Runner for unit tests. I go to Debug -> Run Tests and the content inside the Bracket Tests window is blank. I also can't close that window. It seems like it freezes.
screen shot 2016-10-27 at 9 57 37 am
I was also having this problem before I started editing any files.

Here is my Brackets version:
screen shot 2016-10-27 at 10 00 12 am

And I'm using a 15 inch Macbook Pro and macOS Sierra Version 10.12.1. Please help. Thank you very much.

**Comment by petetnt** [comment]
@arthur801031 Did you run `git submodule update --init` too? 

It might be related to the spec runner window running out of memory and choking too and usually just restarting until it works works too. There's an issue tracking that too somewhere IIRC.

If you have the tests at hand I can try running them too.

**Comment by arthur801031** [comment]
@petetnt I tried restarting the program many times, but Jasmine Spec Runner is still blank. I just submitted the files with the unit test for this feature, but I've never written a unit test before, so I don't know if my code would work. Thank you!

**Comment by arthur801031** [comment]
@petetnt @swmitra 
I figured out how to get Jasmine Spec Runner to run. Here is what I did:

1. I used `npm install chokidar --save` to install chokidar since I was missing the code from chokidar.
2. I used `npm install acorn` to install acorn since I was missing this as well.
3. There were error messages saying that it's looking for 'acorn' in '/Users/i-chun/Documents/brackets/src/extensions/default/JavaScriptCodeHints', so I created a folder named 'node_modules' and put 'acorn' folder inside this directory.
4. There was also an error message saying that it's looking for 'acorn' in '/Users/i-chun/Documents/brackets/test', so I created a folder named 'node_modules' and put 'acorn' folder inside this directory.

Then, all the error messages in Jasmine Spec Runner's DevTools went away, and the Jasmine Spec Runner is now working correctly. 

Note: the unit test that I created for this feature does not work. I will try to modify it. Thank you!

**Comment by zaggino** [comment]
That's weird @arthur801031 , the dependencies you mentioned are already in `package.json` file. Didn't just `npm install` install them for you?

**Comment by arthur801031** [comment]
@zaggino  I think I ran `npm install` before, but I'm not 100% sure. But, in my previous comment's bullet point 3, it seemed that 'HintUtils.js', 'main.js', 'Session.js' were looking for 'node_modules/acorn' in their currernt folder, since that folder was not there the Jasmine Spec Runner crash.

**Comment by arthur801031** [comment]
@petetnt  @swmitra 

The feature and the test function for this feature are now working correctly. I'm not sure why the check has failed for this commit. Please help. Thank you!

File: package.json [hunk 4]
```diff
@@ -1,7 +1,7 @@
 {
     "name": "Brackets",
-    "version": "1.8.0-0",
-    "apiVersion": "1.8.0",
+    "version": "1.9.0-0",
+    "apiVersion": "1.9.0",
```
Comment (petetnt) [inline package.json]: Needs rebase against master so we don't have these duplicate files, `git fetch upstream master && git rebase upstream/master`.

File: src/view/MainViewManager.js [hunk 5]
```diff
@@ -845,6 +846,35 @@ define(function (require, exports, module) {
...
+        if($firstPane.hasClass('active-pane')) {
+            $secondPane.click();
+        }
+        else {
+            $firstPane.click();
+        }
+    }
+
+    /**
+     * Unit test for switching from first pane to second pane
```
Comment (petetnt) [inline src/view/MainViewManager.js]: We can use `SpecRunnerUtils.simulateKeyEvent` instead; for examples see https://github.com/adobe/brackets/blob/f028b3ecdf2017498196f39ddbbe4a2364366c92/test/spec/InstallExtensionDialog-test.js and search for `simulateKeyEvent`

File: src/view/MainViewManager.js [hunk 6]
```diff
@@ -1729,6 +1763,9 @@ define(function (require, exports, module) {
 
     exports.getAllOpenFiles               = getAllOpenFiles;
     exports.focusActivePane               = focusActivePane;
+    exports.switchPaneFocus               = switchPaneFocus;
+    exports.switchPaneUnitTest1To2        = switchPaneUnitTest1To2;
```
Comment (petetnt) [inline src/view/MainViewManager.js]: Not needed, see above

File: src/view/MainViewManager.js [hunk 7]
```diff
@@ -1729,6 +1763,9 @@ define(function (require, exports, module) {
...
+    exports.switchPaneUnitTest2To1        = switchPaneUnitTest2To1;
```
Comment (petetnt) [inline src/view/MainViewManager.js]: Not needed, see above

File: test/spec/MainViewManager-test.js [hunk 8]
```diff
@@ -536,6 +536,31 @@ define(function (require, exports, module) {
                     expect(MainViewManager.getLayoutScheme()).toEqual({rows: 1, columns: 1});
                 });
             });
+            it("should switch pane when alt-w is pressed", function () {
+                runs(function () {
+                    MainViewManager.setLayoutScheme(1, 2);
+                });
+                runs(function () {
+                    MainViewManager.switchPaneUnitTest1To2();
```
Comment (petetnt) [inline test/spec/MainViewManager-test.js]: See above for my comment on how to handle these; the tests are valid but they test if `switchPaneUnitTest1To2` work, instead of testing if the keypress itself work ^^

**Review by petetnt — CHANGES_REQUESTED** [review]
Made some comments about the tests. Feel free to ask extra instructions if needed 👍

File: src/view/MainViewManager.js (hunk 5 above)
Comment (arthur801031) [inline src/view/MainViewManager.js]: Hi @petetnt  I'm having trouble simulating 'Alt-w' key press using SpecRunnerUtils.simulateKeyEvent. Here is my attempt:

`SpecRunnerUtils.simulateKeyEvent(KeyEvent.DOM_VK_ALT, "keydown", $('#first-pane')[0]);
                    SpecRunnerUtils.simulateKeyEvent(KeyEvent.DOM_VK_W, "keydown", $('#first-pane')[0]);
                    SpecRunnerUtils.simulateKeyEvent(KeyEvent.DOM_VK_W, "keypress", $('#first-pane')[0]);
                    SpecRunnerUtils.simulateKeyEvent(KeyEvent.DOM_VK_W, "keyup", $('#first-pane')[0]);`

Could you please let me know what I'm doing wrong? Thank you so much!

Comment (petetnt) [inline src/view/MainViewManager.js]: @arthur801031 Dang, I remembered that SimulateKeyEvent has modifier keys support but it seems not: https://github.com/adobe/brackets/blob/master/test/spec/SpecRunnerUtils.js#L998

Let's modify the tests so that you use `CommandManager.executeCommand` to run `cmd.switchPaneFocus` and check that panes have switches after running that command.

Sorry for the extra trouble. 👍

Comment (arthur801031) [inline src/view/MainViewManager.js]: @petetnt So we don't check whether or not 'Alt-W' is pressed? We only check if cmd.switchPaneFocus can actually switch pane between first pane and second pane.

Comment (petetnt) [inline src/view/MainViewManager.js]: @arthur801031 exactly. I created an issue for adding the modifier key option https://github.com/adobe/brackets/issues/12859

**Comment by arthur801031** [comment]
@petetnt I finished/uploaded the test and rebased against master branch. Please review my code again. Thank you so much! :)

**Review by petetnt — APPROVED** [review]

**Comment by petetnt** [comment]
This LGTM, all tests passing. Merging this, thanks a lot @arthur801031, hopefully you'll contribute to Brackets in the future too!

**Comment by swmitra** [comment]
Great work @arthur801031 and @petetnt in closing this PR 👍

**Comment by schroef** [comment]
SOrry for asking but what button is associated with this, tried google for the icon but cant seem to find it

**Comment by swmitra** [comment]
If I remember correctly it's Ctrl+w. @petent Can you please confirm?

**Comment by petetnt** [comment]
It's Alt+W :+1:

**Comment by schroef** [comment]
PS perhaps you should add a menu item for this. I noticed the new feature but it only tells us its possible now, doesnt say anything about the shortcut. I was actually mixing it up with switching documents, just know i understood that its for switching when 2 panes are active.
